src/annotator/anchoring/pdf.js

```diff
@@ -0,0 +1,430 @@
+    }
+  };
+
+  return getPageTextContent(0).then(count);
+}
+
+/**
+ * Map a position anchor to a region within the text layer of a rendered page.
+ *
+ * If the page is not rendered, a placeholder element is created and the
+ * returned range is the placeholder.
+ *
+ * @param {PDFPageView} page - The PDF.js viewer page
+ * @param {TextPositionAnchor} anchor - Anchor to locate in page
+ * @param {Object} options - Options for `anchor.toSelector`
+ * @return {Range}
+ */
+function anchorByPosition(page, anchor, options) {
+  let renderingDone = false;
+  if (page.textLayer) {
+    renderingDone = page.textLayer.renderingDone;
+  }
+  if (page.renderingState === RenderingStates.FINISHED && renderingDone) {
+    // The page has been rendered. Use HTML anchoring to locate the quote in
+    // the text layer.
+    const root = page.textLayer.textLayerDiv;
+    const selector = anchor.toSelector(options);
+    return html.anchor(root, [selector]);
+  }
+
+  // The page has not been rendered yet. Create a placeholder element and
+  // anchor to that instead.
+  const div = page.div || page.el;
+  let placeholder = div.getElementsByClassName('annotator-placeholder')[0];
+  if (!placeholder) {
+    placeholder = document.createElement('span');
+    placeholder.classList.add('annotator-placeholder');
+    placeholder.textContent = 'Loading annotations…';
+    div.appendChild(placeholder);
+  }
+  const range = document.createRange();
+  range.setStartBefore(placeholder);
+  range.setEndAfter(placeholder);
+  return range;
+}
+
+/**
+ * Search for a quote in the given pages.
+ *
+ * @param {number[]} pageIndexes - Pages to search in priority order
+ * @param {TextQuoteSelector} quoteSelector
+ * @return {Promise<Range>} Location of quote
+ */
+function findInPages(pageIndexes, quoteSelector, positionHint) {
+  if (pageIndexes.length === 0) {
+    return Promise.reject(new Error('Quote not found'));
+  }
+
+  const [pageIndex, ...rest] = pageIndexes;
+
+  const content = getPageTextContent(pageIndex);
+  const page = getPage(pageIndex);
+  const offset = getPageOffset(pageIndex);
+
+  const attempt = ([, content, offset]) => {
+    const root = { textContent: content };
+    const anchor = TextQuoteAnchor.fromSelector(root, quoteSelector);
+    if (positionHint) {
+      let hint = positionHint.start - offset;
+      hint = Math.max(0, hint);
+      hint = Math.min(hint, content.length);
+      return anchor.toPositionAnchor({ hint });
+    }
+    return anchor.toPositionAnchor();
+  };
+
+  const next = () => findInPages(rest, quoteSelector, positionHint);
+
+  const cacheAndFinish = anchor => {
+    if (positionHint) {
+      if (!quotePositionCache[quoteSelector.exact]) {
+        quotePositionCache[quoteSelector.exact] = {};
+      }
+      quotePositionCache[quoteSelector.exact][positionHint.start] = {
+        page,
+        anchor,
+      };
+    }
+    return anchorByPosition(page, anchor);
+  };
+
+  return Promise.all([page, content, offset])
```

Comment:
This function is not simple, but this promise chain is relatively easy to follow.

Reply:
I've added some extra comments to hopefully clarify the overall flow.